### ISSUE TITLE
Unknown hang on Travis at termination of MRI stdlib suite

### ISSUE THREAD

**Issue by headius** [jruby/jruby#5718]
Something in the MRI stdlib test suite (`rake test:mri:stdlib`) appears to hang when shutting down the test suite. The suite itself runs green, but after the last test completes it appears to intermittently hang and fail to complete, which leads to Travis marking it as a failure.

In 44d0e0c26bec89b99f214497465b927447491e64 I have moved this suite to allowed failures, but we will want to figure out why this suite sometimes hangs and move it back out.

**Comment by ahorek** [comment]
Hi @headius, I had a similar issue with https://github.com/rails/sprockets

so I've decided to investigate a little and here's a minimum repro
https://github.com/ahorek/jrubyhang

the test passes, but on jruby it hangs during shutdown.

**Comment by kares** [comment]
@ahorek you seem to be using `Concurrent` library, which isn't used by minitest in the MRI suite, right?
... your Promise will likely spin-up an executor, which might need some proper (manual) shutdown.
wonder if that worked in previous versions, anyway this sounds like an `at_exit` pool shutdown issue.

**Comment by ahorek** [comment]
> you seem to be using Concurrent library, which isn't used by minitest in the MRI suite, right?

right, but I think it might be the same problem.

> which might need some proper (manual) shutdown

should Concurrent::Promise workers be autoterminated? I didn't find a documentation about how to shutdown these workers manually in the code.
This issue might be also relevant https://github.com/ruby-concurrency/concurrent-ruby/issues/701 @pitr-ch ?
the same test passes on MRI.

it looks like it's waiting for other threads to terminate during minitests shutdown. But it seems to terminate properly in irb.

**Comment by headius** [comment]
Aha, are those executor threads being created as daemon threads? If not, they'd keep the JVM alive until (I believe) some timeout in the executor actually shuts them down. Or the executor could be eagerly shut down at some point.

**Comment by headius** [comment]
So at least for the concurrent-ruby case, it does appear to be a thread pool created with non-daemon threads that never shut down when the Ruby app completes.

I see a stack trace like this (note it is not marked as a daemon thread):

```
"pool-1-thread-1" #16 prio=5 os_prio=31 tid=0x00007f916054c000 nid=0x5803 waiting on condition [0x000070000eff6000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000076de77188> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

And if I dig into the heap to see who's holding references to that thread, I eventually get back to a thread pool held by:

```
varTable in com.concurrent_ruby.ext.SynchronizationLibrary$JRubyLockableObject#2
```

So that issue does appear to be a problem with the promise-running thread pool in concurrent-ruby not being configured with daemon threads. Thoughts on this, @petr-ch?

I do not know if this relates to the JRuby test suite hang, since I would not expect anything in that suite to use concurrent-ruby.

**Comment by ahorek** [comment]
@headius this looks like to be the cause https://github.com/jruby/jruby/issues/5936 for the original problem, take a look

**Comment by headius** [comment]
It turns out the problem was reported elsewhere. The hanging stdlib suite appears to be working properly again after #5942 due to the fix for #5863, so it ended up being a regression after all. This is now fixed for 9.2.9.
